idaes/power_generation/carbon_capture/compression_system/compressor.py

```diff
@@ -0,0 +1,603 @@
+        def Ma_con(b, t):
+            # not sure if there are two phases at the inlet
+            # but I assume that this contains single vapor phase.
+            speed_of_sound = self.control_volume.properties_in[
+                t].speed_sound_phase['Vap']
+            return b.Ma[t] == b.U2[t] / speed_of_sound
+
+        @self.Constraint(self.flowsheet().config.time, doc='Pressure Ratio')
+        def Pratio_con(b, t):
+            return properties_out[t].pressure == b.ratioP[t] * \
+                  properties_in[t].pressure
+
+        @self.Constraint(self.flowsheet().config.time, doc="Rotation Speed "
+                         "of the Impeller")
+        def rspeed_con(b, t):
+            return b.U2[t] == 2 * const.pi * b.r2 * b.rspeed[t]
+
+        # set up the vane diffuser rule.
+        vdselect = self.config.vane_diffuser_type
+        if vdselect is not VaneDiffuserType.custom:
+            _log.warning("A vane diffuser callback was provided"
+                         "but the valve diffuser type is not custom.")
+        if vdselect == VaneDiffuserType.vane_diffuser:
+            _build_vane_diffuser_equations(self)
+        elif vdselect == VaneDiffuserType.vaneless_diffuser:
+            _build_vaneless_diffuser_equations(self)
+
+        # set up the impeller rule.
+        iselect = self.config.impeller_type
+        if iselect is not ImpellerType.custom:
+            _log.warning("An impeller callback was provided but the impeller"
+                         "type is not custom.")
+        if iselect == ImpellerType.cover_impeller:
+            _build_cover_impeller_equations(self)
+        elif iselect == ImpellerType.open_impeller:
+            _build_open_impeller_equations(self)
+
+        @self.Constraint(self.flowsheet().config.time, doc="Polytropic "
+                         "Efficiency")
+        def eff_p_v_cons(b, t):
+            return b.eff_p_v[t] * b.impeller_work_coeff[t] == b.mu_p_v[t]
+
+        # --------------------------------------------------------------------
+
+        # Calculate total enthalpy and entropy change through the stage
+        @self.Expression(self.flowsheet().config.time, doc="Specific "
+                         "Enthalpy Change of Isentropic Process")
+        def delta_enth_isentropic(b, t):
+            return b.properties_isentropic[t].enth_mol - \
+                properties_in[t].enth_mol
+
+        @self.Expression(self.flowsheet().config.time, doc="Actual Enthalpy "
+                         "Change")
+        def delta_enth_actual(b, t):
+            return properties_out[t].enth_mol - properties_in[t].enth_mol
+
+        @self.Expression(self.flowsheet().config.time, doc="Entropy change in"
+                         "Compressor Stage")
+        def deltaS(b, t):
+            return properties_out[t].entr_mol - properties_in[t].entr_mol
+
+        @self.Constraint(self.flowsheet().config.time, doc="Isentropic"
```

Comment:
I was having some trouble with convergence and how the initialization worked, and going through the model in more detail, and I found that it seem like the isentropic  efficiency constraint seems to be duplicated here and in the pressure changer model.

Reply:
hi @eslickj : I comment out those lines and they give me an extra degree of freedom when I run the test. I try to fix the impeller tip radius variable (r2) but the solver throw me a below error.
pyutilib.common._exceptions.ApplicationError: Solver (ipopt) did not exit normally

Reply:
The extra degree of freedom could be a result of a missing constraint due to the earlier missing constraint decorator.

Reply:
Hi @eslickj : Thanks. I check the model and adding the missing earlier decorator. Now I still get the same error. Please see the below.
c:\users\minh2\miniconda3\envs\compression\lib\site-packages\pyomo\opt\base\solvers.py:575: in solve
    self._presolve(*args, **kwds)
c:\users\minh2\miniconda3\envs\compression\lib\site-packages\pyomo\opt\solver\shellcmd.py:198: in _presolve
    OptSolver._presolve(self, *args, **kwds)
c:\users\minh2\miniconda3\envs\compression\lib\site-packages\pyomo\opt\base\solvers.py:675: in _presolve
    **kwds)
c:\users\minh2\miniconda3\envs\compression\lib\site-packages\pyomo\opt\base\solvers.py:746: in _convert_problem
    **kwds)
c:\users\minh2\miniconda3\envs\compression\lib\site-packages\pyomo\opt\base\convert.py:105: in convert_problem
    problem_files, symbol_map = converter.apply(*tmp, **tmpkw)
c:\users\minh2\miniconda3\envs\compression\lib\site-packages\pyomo\solvers\plugins\converter\model.py:191: in apply
    io_options=io_options)
c:\users\minh2\miniconda3\envs\compression\lib\site-packages\pyomo\core\base\block.py:1814: in write
    io_options)
pyomo\repn\plugins\ampl\ampl_.pyx:362: in pyomo.repn.plugins.ampl.ampl_.ProblemWriter_nl.__call__
    ???
pyomo\repn\plugins\ampl\ampl_.pyx:363: in pyomo.repn.plugins.ampl.ampl_.ProblemWriter_nl.__call__
    ???
pyomo\repn\plugins\ampl\ampl_.pyx:365: in pyomo.repn.plugins.ampl.ampl_.ProblemWriter_nl.__call__
    ???
pyomo\repn\plugins\ampl\ampl_.pyx:958: in pyomo.repn.plugins.ampl.ampl_.ProblemWriter_nl._print_model_NL
    ???
pyomo\repn\standard_repn.pyx:372: in pyomo.repn.standard_repn.generate_standard_repn
    ???
pyomo\repn\standard_repn.pyx:978: in pyomo.repn.standard_repn._generate_standard_repn
    ???
pyomo\repn\standard_repn.pyx:477: in pyomo.repn.standard_repn._collect_sum
    ???
pyomo\repn\standard_repn.pyx:950: in pyomo.repn.standard_repn._collect_standard_repn
    ???
pyomo\repn\standard_repn.pyx:843: in pyomo.repn.standard_repn._collect_negation
    ???
pyomo\repn\standard_repn.pyx:950: in pyomo.repn.standard_repn._collect_standard_repn
    ???
pyomo\repn\standard_repn.pyx:477: in pyomo.repn.standard_repn._collect_sum
    ???
pyomo\repn\standard_repn.pyx:950: in pyomo.repn.standard_repn._collect_standard_repn
    ???
pyomo\repn\standard_repn.pyx:862: in pyomo.repn.standard_repn._collect_identity
    ???
pyomo\repn\standard_repn.pyx:950: in pyomo.repn.standard_repn._collect_standard_repn
    ???
pyomo\repn\standard_repn.pyx:570: in pyomo.repn.standard_repn._collect_prod
    ???
pyomo\repn\standard_repn.pyx:950: in pyomo.repn.standard_repn._collect_standard_repn
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   TypeError: unsupported operand type(s) for *: 'int' and 'dict'

pyomo\repn\standard_repn.pyx:656: TypeError